compiler/rustc_mir_transform/src/check_finalizers.rs

```diff
@@ -70,18 +48,186 @@ impl<'tcx> MirPass<'tcx> for CheckFinalizers {
+    arg: Span,
+    tcx: TyCtxt<'tcx>,
+    param_env: ParamEnv<'tcx>,
+}
+
+impl<'tcx> FinalizationCtxt<'tcx> {
+    fn check(&mut self, ty: Ty<'tcx>) {
+        if self.is_finalizer_safe(ty) || self.is_no_finalize(ty) {
+            return;
+        }
+
+        // We must now recurse through the `Ty`'s component types and search for
+        // all the `Drop` impls. If we find any, we have to check that there are
+        // no unsound projections into fields in their drop method body. More
+        // specifically: if one of the drop methods dereferences a field which
+        // is !Send, !Sync, or !FinalizerSafe, we must throw an error.
+        match ty.kind() {
+            ty::Infer(ty::FreshIntTy(_))
+            | ty::Infer(ty::FreshFloatTy(_))
+            | ty::Bool
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Never
+            | ty::FnDef(..)
+            | ty::FnPtr(_)
+            | ty::Char
+            | ty::GeneratorWitness(..)
+            | ty::RawPtr(_)
+            | ty::Ref(..)
+            | ty::Str
+            | ty::Foreign(..) => {
+                // None of these types can implement Drop.
+                return;
+            }
+            ty::Dynamic(..) | ty::Error(..) => {
+                // Dropping a trait object uses a virtual call, so we can't
+                // work out which drop method to look at compile-time. This
+                // means we must be more conservative and bail with an error
+                // here, even if the drop impl itself would have been safe.
+                self.emit_err();
+            }
+            ty::Slice(ty) => self.check(*ty),
+            ty::Array(elem_ty, ..) => {
+                self.check(*elem_ty);
+            }
+            ty::Tuple(fields) => {
+                // Each tuple field must be individually checked for a `Drop`
+                // impl.
+                fields.iter().for_each(|f_ty| self.check(f_ty));
+            }
+            ty::Adt(def, substs) if !self.is_copy(ty) => {
+                if def.has_dtor(self.tcx) {
+                    if def.is_box() {
+                        // This is a special case because Box has an empty drop
+                        // method which is filled in later by the compiler.
+                        self.emit_err();
+                    }
+
+                    let drop_trait = self.tcx.require_lang_item(LangItem::Drop, None);
+                    let drop_fn = self.tcx.associated_item_def_ids(drop_trait)[0];
+                    let substs = self.tcx.mk_substs_trait(ty, substs);
+                    let instance = ty::Instance::resolve(self.tcx, self.param_env, drop_fn, substs)
+                        .unwrap()
+                        .unwrap();
+                    let mir = self.tcx.instance_mir(instance.def);
+                    let mut checker = ProjectionChecker { cx: self, body: mir };
+                    checker.visit_body(&mir);
```

Comment:
I'm not familiar with rustc's innards: is this the bit which (indirectly) does the "call `visit_projection`"?

Reply:
That's right, this kicks off the `MirVisitor` which walks the CFG and then executes our code when it sees a projection.

Reply:
Got it!